File: lambda/redirect.py
import json
import logging
import os

import boto3
from botocore.exceptions import ClientError

logger = logging.getLogger()
logger.setLevel(logging.INFO)


class Result:
    def __init__(self, status, body=None, base64=False, headers=None, cookies=None):
        self.status = status
        if body:
            self.body = body
        else:
            self.body = None
        self.base64 = base64
        if not headers:
            self.headers = {}
        else:
            self.headers = dict(headers)
        if not cookies:
            self.cookies = []
        else:
            self.cookies = list(cookies)
    
    def format_api_gateway(self):
        result = {
            'statusCode': self.status,
            'isBase64Encoded': self.base64,
        }
        if self.body:
            result['body'] = self.body
        if self.headers:
            result['headers'] = self.headers
        if self.cookies:
            result['cookies'] = self.cookies
        return result


class ErrorResult(Result):
    def __init__(self, status, message):
        body = {'errorMessage': message}
        super().__init__(status, json.dumps(body))


class RedirectResult(Result):
    def __init__(self, location):
        headers = {'Location': location}
        super().__init__(301, None, headers=headers)


def query(table, host):
    dynamodb = boto3.client('dynamodb')
    key={'host': {'S': host}}
    try:
        result = dynamodb.get_item(TableName=table, Key=key)
    except ClientError:
        return ErrorResult(500, "Unable to query for redirect")

    if location := result.get('Item', {}).get('location', {}).get('S', None):
        return RedirectResult(location)

    return ErrorResult(404, f"No destination for {host}")


def lambda_handler(event, context):
    host = event['headers']['host']
    logger.debug("Requested host: %s", host)
    table_name = os.environ.get('DYNAMODB_TABLE', 'redirection')
    logger.debug("Querying from DynamoDB table: %s", table_name)
    response = query(table_name, host).format_api_gateway()
    logger.debug("Response: %s", json.dumps(response))
    return response

File: lambda/redirect.ts
import { APIGatewayProxyEventV2, APIGatewayProxyResultV2, Context } from "aws-lambda";
import { DynamoDB } from "@aws-sdk/client-dynamodb";
import { DynamoDBDocument } from "@aws-sdk/lib-dynamodb";
import { Tracer } from "@aws-lambda-powertools/tracer";
import { Logger } from "@aws-lambda-powertools/logger";

const logger = new Logger({ serviceName: "redirectHandler" });
const tracer = new Tracer({ serviceName: "redirectHandler" });
tracer.provider.setLogger(logger);

const dynamodbRaw = tracer.captureAWSv3Client(new DynamoDB({ useFipsEndpoint: true }));
const dynamodb = tracer.captureAWSv3Client(DynamoDBDocument.from(dynamodbRaw));

const table = process.env.REDIRECT_TABLE!;

const unknownHost: APIGatewayProxyResultV2 = {
  statusCode: 404,
  body: "No redirect location was found for the given host",
};
const badRequest: APIGatewayProxyResultV2 = {
  statusCode: 400,
  body: "The request was invalid",
};
const unexpectedError: APIGatewayProxyResultV2 = {
  statusCode: 500,
  body: "An unexpected error occurred",
};

export async function handler(event: APIGatewayProxyEventV2, context: Context): Promise<APIGatewayProxyResultV2> {
  logger.addContext(context);
  if (logger.isColdStart()) {
    console.info("This event is a cold start");
  }

  const host = event.headers.host;
  if (!host) {
    logger.error("Invalid request received", { data: event.headers });
    return badRequest;
  }

  try {
    const record = await dynamodb.get({ TableName: table, Key: { host }, AttributesToGet: ["location"] });
    if (!record.Item) {
      logger.error("Unknown host", { data: event.headers });
      return unknownHost;
    }
    if (!record.Item.location) {
      logger.error("Unexpected record format", { data: record.Item });
      return unexpectedError;
    }
    return {
      statusCode: 301,
      headers: {
        Location: record.Item.location,
      },
    };
  } catch (err) {
    logger.error("Unexpected DynamoDB service error", err as Error);
    return unexpectedError;
  }
}
